Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
+      } else {
+        var g = isGhost[dims - 1];
+        return Tuple.Create(g, MakeTupleKey(isGhost, dims - 1));
+      }
+    }
+
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> ghostArguments = null) {
       Contract.Requires(tok != null);
       Contract.Requires(0 <= dims);
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
+        ghostArguments = new bool[dims].ToList().ConvertAll(_ => false);
+      }
+      object key = MakeTupleKey(ghostArguments, dims);
       if (!tupleTypeDecls.TryGetValue(dims, out tt)) {
         Contract.Assume(allowCreationOfNewType);  // the parser should ensure that all needed tuple types exist by the time of resolution
-        if (dims == 2) {
+        if (dims == 2 && !ghostArguments.Contains(true)) {
           // tuple#2 is already defined in DafnyRuntime.cs
-          tt = new TupleTypeDecl(dims, SystemModule, DontCompile());
+          tt = new TupleTypeDecl(ghostArguments, SystemModule, DontCompile());
         } else {
-          tt = new TupleTypeDecl(dims, SystemModule, null);
+          tt = new TupleTypeDecl(ghostArguments, SystemModule, null);
         }
         tupleTypeDecls.Add(dims, tt);
         SystemModule.TopLevelDecls.Add(tt);
       }
       return tt;
     }
 
-    public static string TupleTypeName(int dims) {
-      Contract.Requires(0 <= dims);
-      return "_tuple#" + dims;
+    public static char IsGhostToChar(bool isGhost) {
+      return isGhost ? 'G' : 'O';
+    }
+
+    public static bool IsGhostFromChar(char c) {
+      Contract.Requires(c == 'G' || c == 'O');
+      return c == 'G';
+    }
+
+    public static string GhostArgumentsToString(List<bool> ghostArguments) {
```

Comment:
Is this string user-facing? We could make it more or less complex. Less complex would be to always return `String.Concat(ghostArguments.ConvertAll(IsGhostToChar)))` even if there are no ghosts, and more complex would be to return `3G2O` instead of `5GGGOO`.

This looks good to me though, just wanted to mention some alternatives.